rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,221 @@
+
+This approach was rejected because:
+
+- The API calls made by each app are [stubbed at the request level](https://github.com/alphagov/gds-api-adapters/tree/8a67c1e2d65ac46533725730fa68d24485c9d964/lib/gds_api/test_helpers), so there is little benefit in re-testing the associated functionality if our API adapters are guaranteed by contract tests.
+
+- The environment we have for sandboxed E2E tests does not reflect our real environments [[1](https://github.com/alphagov/publishing-e2e-tests/pull/386)] [[2](https://github.com/alphagov/publishing-e2e-tests/pull/383)] [[3](https://github.com/alphagov/publishing-e2e-tests/pull/379)] [[4](https://github.com/alphagov/publishing-e2e-tests/pull/367)] [[5](https://github.com/alphagov/publishing-e2e-tests/pull/326)]. It's unclear if something that works in this environment will also work in a real one.
+
+- The environment we have for sandboxed E2E tests is notoriously slow, brittle and complex [[1](https://github.com/alphagov/publishing-e2e-tests/pull/381)] [[2](https://github.com/alphagov/publishing-e2e-tests/pull/360)] [[3](https://github.com/alphagov/publishing-e2e-tests/pull/355)] [[4](https://github.com/alphagov/publishing-e2e-tests/pull/331)] [[5](https://github.com/alphagov/publishing-e2e-tests/pull/284)]. We should invest in more granular forms of testing instead of this one [[1](https://github.com/alphagov/publishing-e2e-tests/blob/f12ce7a9626b38728758cf157ea883907bd89de2/docs/what-belongs-in-these-tests.md)] [[2](https://technology.blog.gov.uk/2019/01/29/lessons-learnt-using-contract-testing-in-gov-uk-pay/)].
+
+### Check: detailed features of GOV.UK work
+
+A previous draft of this RFC would have required a detailed audit of the functionality of each app, to ensure we had Smoke or sandboxed E2E tests for each of its features, for some definition of "feature".
+
+This approach was rejected because:
+
+- There is no clear way to demarcate features of an app, let alone GOV.UK as a whole. Given the sheer volume of functionality in our apps, it's not practical to audit or have realistic tests for all of it.
+
+- While we have existing tests for specific features of GOV.UK [[1](https://github.com/alphagov/publishing-e2e-tests/tree/b4a7961cdcd200623a694baf833724a41c98d051/spec/collections_publisher)] [[2](https://github.com/alphagov/smokey/blob/a60df2cde5c886fa6d54e8cd820d90facca89e8e/features/csv_preview.feature)], there is no thorough, positive logic to justify their existence or extension [[1](https://github.com/alphagov/govuk-rfcs/blob/master/rfc-050-do-end-to-end-testing-of-gov-uk-applications.md#solution)] [[2](https://github.com/alphagov/publishing-e2e-tests/blob/f12ce7a9626b38728758cf157ea883907bd89de2/docs/what-belongs-in-these-tests.md)]. This is a weak foundation to try and build upon.
+
+## Consequences
+
```

Comment:
I think one aspect not covered in this RFC and that I think should be a consequence of continuous deployment, is that best practice needs to be for Pull Requests to be small. Large PRs which make multiple changes, are the most likely to cause problems, and the most difficult to rollback or fix. In general, a large number of PRs is better than a few large PRs. Small PR tend to be the least risky to merge (or where risks exist they tend to be better understood) and thereby mitigate some of the risks added by moving to continuous deployment.

Reply:
+1 to that.

I do think it's slightly out-of-scope for this RFC to try and mandate how people write their PRs. For one thing, it's difficult to measure "size", and there can be reasonable exceptions to a rule.

I could still mention this as part of [being more cautious](https://github.com/alphagov/govuk-rfcs/blob/continuous-deployment/rfc-128-continuous-deployment.md#being-more-cautious). So we not only need to rethink our muscle memory for clicking merge, but also how we got there in the first place, if you follow me.

How does that sound?

Reply:
We do have some [styleguides around pull requests](https://github.com/alphagov/styleguides/blob/master/pull-requests.md), although this is probably superseded by the [GDS Way](https://gds-way.cloudapps.digital/standards/pull-requests.html#pull-requests). Perhaps the GDS way is a good place to think about calling out small changes.

Reply:
I'm going to resolve this thread for now. @reggieb feel free to reopen if you have more thoughts about how our PR practices should feature in this RFC.

Reply:
:+1: